include/deal.II/base/symmetric_tensor.h

```diff
@@ -3544,6 +3544,121 @@ DEAL_II_HOST constexpr inline DEAL_II_ALWAYS_INLINE
   return result;
 }
 
+/**
+ * An enumeration to describe the desired symmetry of the
+ * symmetric fourth order tensor
+ */
+enum class SymmetryType
+{
+  /**
+   * Only minor symmetry of the form:
+   * $A_{ijkl}=A_{jikl}=A_{ijlk}=A_{jilk}$.
+   */
+  minor,
+  /** 
+   * Both minor and major symmetry of the form:
+   * Minor - $A_{ijkl}=A_{jikl}=A_{ijlk}=A_{jilk}$
+   * Major - $A_{ijkl}=A_{klij}$.
+   */
+  minor_major,
+};
+
+/**
+ * Return the symmetrized version of a full rank-4 tensor, i.e.
+ * as a symmetric rank-4 tensor.
+ * @param symmetry_type decides the type of symmetry.
+ * This is the version for general dimensions.
+ *
+ * @relatesalso SymmetricTensor
+ */
+template <int dim, typename Number>
+DEAL_II_HOST constexpr inline DEAL_II_ALWAYS_INLINE
+  SymmetricTensor<4, dim, Number>
+  symmetrize(const Tensor<4, dim, Number> &t,
+             const SymmetryType &symmetry_type = SymmetryType::minor)
+{
+  SymmetricTensor<4, dim, Number> result;
+  for (unsigned int i = 0; i < dim; ++i)
+    result[i][i][i][i] = t[i][i][i][i];
+
+  const Number half = internal::NumberType<Number>::value(0.5);
+
+  if (symmetry_type == SymmetryType::minor)
```

Comment:
This would lend itself to a `switch`-`case` statement. In any case, you should add a `default` case (or, in the current `if-else if` structure, a final `else`) that simply says `DEAL_II_NOT_IMPLEMENTED()` to capture the case where someone passes in an unsupported case for the second argument.